README.md
# Upgrade Unattach and Re-attach Media Attachments

WordPress allows to attach (or assign) images and other attachments organized within the Media Library to posts and pages. A file will automatically be attached to a particular post when uploaded using the Add media button in that post. There are also "Attach" links in the Media Library to manually attach files to posts. Once attached there is no way to unattach or re-attach images, pdfs or any attachment in WordPress core.

This plugin fills this gap by providing two additional links next to "Attach" in the **Media Library**:

*   **Re-Attach**: Allows to choose a new post to attach the file to
*   **Unattach**: Totally removes an existing relation between an attachment and a post

## Installation requirements
* Requires WP at least: 4.9
* Tested up to: 5.4
* Stable tag: 1.2.0

## Bulk Actions
The plugin also provides bulk actions for unattaching and re-attaching multiple files at once! Just mark the files you want to change, select your desired action from the *Bulk Actions* dropdown and hit *Apply*.

## Installation
1. Upload the plugin to the `/wp-content/plugins/` directory or from the repository.
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Refer to plugin description in regards to setting up how the plugins works
1. **It will work out of the box**. Move to your media library and you can unattach and re-attach attachments to various posts.

## Note
[My initial issue came from here](https://stackoverflow.com/questions/45690582/image-post-attachments-not-specific-to-post/45705413#45705413). Thanks to [Unattach and Re-attach Media Attachments](https://wordpress.org/plugins/unattach-and-re-attach-attachments/) which inspired this plugin. Just adding upgrades of the code to the latter so it is safer for use and have fixed some bugs.

## Frequently Asked Questions
1. How can I contribute?
* You can raise lots of issues here and also make some [Pull Requests through github](https://github.com/bahiirwa/upgrade-for-unattach-re-attach-media-attachments/)

## Screenshots
1. Bulk actions "Unattach" and "Re-Attach" in the Media Library
1. The extended "Parent" column in the Media Library
1. Individual actions "Unattach" and "Re-Attach" per post

## Changelog

### Version 1.2.0 (22.04.2020)
1. Improve Internalization for plugin.
1. Improve file structure for better code management.
1. Structured the code for WordPress code Standards.
1. Text for WP 5.4.

### Version 1.1.1 (15.11.2017) =
1. Fixed issue/bug with translatability
1. Gutenberg Tested - New Editor WP 5.0.0

### Version 1.1.0 (15.10.2017)
* Fixed issue/bug with attach and unattach options appearing on all bulk action select tags.
* Tested for WP Version 4.8.2
* Better Documentation

### Version 1.0.0 (16.08.2017)
* Security Upgrades added the inspire plugin
* Refactor the code to make it leaner
* Easy to read code.


changelog.txt
= Version 1.1.1 (15.11.2017) =
1. Fixed issue/bug with translatability
1. Gutenberg Tested - New Editor WP 5.0.0

= Version 1.1.0 (15.10.2017) =
1. Fixed issue/bug with attach and unattach options appearing on all bulk action select tags.
1. Tested for WP Version 4.8.2
1. Better Documentation

= Version 1.0.0 (16.08.2017) =
1. Security Upgrades added the inspire plugin
1. Refactor the code to make it leaner
1. Easy to read code.


lurma.php
<?php
/**
 * Plugin Name: Upgrade for Unattach & Re-Attach Media Attachments
 * Plugin URI: https://github.com/bahiirwa/Upgrade-for-Unattach-Re-Attach-Media-Attachments/
 * Description: Secure, powerful but simple plugin to Unattach and Re-attach images and other attachments from within the media library.
 * Version: 1.0.1
 * Author: Laurence Bahiirwa
 * Author URI: https://omukiga.com
 * Requires at least: 3.0
 * Tested up to: 4.8.1
 * Tags: Lurma, Attachments, Unattach, Re-Attach, Image, Media, Library, Detach, Assign
 * Text Domain: lurma
 * License: GPLv2 or Later
**/

/*
 * @package   lurma
 * @link      https://github.com/bahiirwa/lurma
 * Basic Security: Exit plugin if accessed directly.
 */
 if ( ! defined( 'ABSPATH' ) ) {
	echo 'Hi there! You are up to no good!';
	exit;
 }

 define( 'LURMA_PLUGIN_DIR', plugin_dir_path( __FILE__ ) );

/*
 * Plugin Callback Functions
 */
 require_once( LURMA_PLUGIN_DIR . '/includes/functions.php' );

/*
 * Run Plugin Functions
*/
 add_action( 'admin_menu', 'lurma_admin_menu' );
 add_filter( 'manage_upload_columns', 'lurma_manage_upload_columns') ;
 add_action( 'manage_media_custom_column', 'lurma_manage_media_custom_column', 0, 2 );

 if( is_admin() ) {
	add_action( 'admin_footer', 'lurma_custom_bulk_admin_footer' );
	add_action( 'load-upload.php', 'lurma_custom_bulk_action' );
 }


includes/bulk-functions.php
<?php
/**
 * Basic security.
 *
 * @package Lurma.
 */

defined( 'ABSPATH' ) || exit;

/**
 * Adds new buld actions 'unattach' and 're-attach' to the media lib.
 *
 * @see http://wordpress.stackexchange.com/questions/29822/custom-bulk-action
 * @return void
 */
function lurma_custom_bulk_admin_footer() {
	global $post_type;
	if ( is_admin() ) {
		?>
			<script type="text/javascript">
				jQuery(document).ready(function() {
					$ = jQuery;
					$( '<option>' ).val( 'unattach' ).text( '<?php esc_attr_e( 'Unattach', 'lurma' ); ?> ').appendTo("select[name='action']");
					$( '<option>' ).val( 'reattach' ).text( '<?php esc_attr_e( 'Re-Attach', 'lurma' ); ?> ').appendTo("select[name='action']");
					$( '<option>' ).val( 'unattach' ).text( '<?php esc_attr_e( 'Unattach', 'lurma' ); ?> ').appendTo("select[name='action2']");
					$( '<option>' ).val( 'reattach' ).text( '<?php esc_attr_e( 'Re-Attach', 'lurma' ); ?> ').appendTo("select[name='action2']");

					$( '#doaction, #doaction2' ).click(function(e){
						$( 'select[name^="action"]' ).each(function(){
							if ( $(this).val() == '<?php esc_attr_e( 'reattach', 'lurma' ); ?>' ) {
								e.preventDefault();
								findPosts.open();
							}
						});
					});
				});
			</script>
		<?php
	}
}

/**
 * Implements a bulk action for unattaching items in bulk.
 *
 * @see http://wordpress.stackexchange.com/questions/91874/how-to-make-custom-bulk-actions-work-on-the-media-upload-page
 * @see http://www.skyverge.com/blog/add-custom-bulk-action/
 * @return void
 */
function lurma_custom_bulk_action() {

	// If ($post_type == 'attachment' ) {  REPLACE WITH:.
	if ( ! isset( $_REQUEST['detached'] ) ) {

		// Get the action.
		$wp_list_table = _get_list_table( 'WP_Media_List_Table' );
		$action        = $wp_list_table->current_action();

		$allowed_actions = array( 'unattach' );

		if ( ! in_array( $action, $allowed_actions, true ) ) {
			return;
		}

		// Security check.
		// Check_admin_referer( 'bulk-posts' ); REPLACE WITH:.
		check_admin_referer( 'bulk-media' );

		// Make sure ids are submitted.  depending on the resource type, this may be 'media' or 'ids'.
		if ( isset( $_REQUEST['media'] ) ) {
			$post_ids = array_map( 'intval', $_REQUEST['media'] );
		}

		if ( empty( $post_ids ) ) {
			return;
		}

		// This is based on wp-admin/edit.php.
		$sendback = remove_query_arg( array( 'unattached', 'untrashed', 'deleted', 'ids' ), wp_get_referer() );

		if ( ! $sendback ) {
			$sendback = admin_url( "upload.php?post_type=$post_type" );
		}

		$pagenum  = $wp_list_table->get_pagenum();
		$sendback = add_query_arg( 'paged', $pagenum, $sendback );

		switch ( $action ) {
			case 'unattach':
				global $wpdb;

				// If we set up user permissions/capabilities, the code might look like:.
				// If ( !current_user_can($post_type_object->cap->export_post, $post_id) ).
				// wp_die( __( 'You are not allowed to unattach this post.' ) );.
				if ( ! is_admin() ) {
					wp_die( esc_attr__( 'You are not allowed to unattach this post.', 'lurma' ) );
				}

				$unattached = 0;
				foreach ( $post_ids as $post_id ) {
					// Alter post to unattach media file.
					if (
						false === $wpdb->update( $wpdb->posts,
							array(
								'post_parent' => 0,
							),
							array(
								'id'        => (int) $post_id,
								'post_type' => 'attachment',
							)
						)
					) {
						wp_die( esc_attr__( 'Error unattaching post.', 'lurma' ) );
					}
					$unattached++;
				}

				$sendback = add_query_arg(
					array(
						'unattached' => $unattached,
						'ids'        => join( ',', $post_ids ),
					),
					$sendback
				);

				break;
			default: return;
		}

		$sendback = remove_query_arg( array( 'action', 'action2', 'tags_input', 'post_author', 'comment_status', 'ping_status', '_status', 'post', 'bulk_edit', 'post_view' ), $sendback );

		wp_safe_redirect( $sendback );
		exit();

	}

}


includes/manage-columns.php
<?php
/**
 * Plugin Functions.
 * Basic security.
 *
 * @package Lurma.
 */

defined( 'ABSPATH' ) || exit;

/**
 * Implements filter 'manage_upload_columns' to replace column 'parent' without custom column 'extended_parent'.
 *
 * @param array $columns Passed Array of columns from media posts.
 * @return array $columns New Array of columns from media posts.
 */
function lurma_manage_upload_columns( $columns ) {
	unset( $columns['parent'] );
	$columns['extended_parent'] = esc_attr__( 'Parent Post', 'lurma' );
	return $columns;
}

/**
 * Implementes action 'manage_media_custom_column' to add a column into the media page with link Attach, Unattach, Re-Attach.
 *
 * @param array   $column_name Column from media posts.
 * @param integer $id          ID for the attachment post.
 * @return void
 */
function lurma_manage_media_custom_column( $column_name, $id ) {

	$post = get_post( $id );

	// Only act on our custom column extended_parent.
	if ( 'extended_parent' !== $column_name ) {
		return;
	}

	$post_parent = $post->post_parent;

	if ( $post_parent ) {

		if ( get_post( $post_parent ) ) {
			$title = _draft_or_post_title( $post_parent );
		}

		$url_unattach = admin_url( 'tools.php?page=unattach&noheader=true&post_id=' . $post->ID );

		?>
			<div>
				<strong><a href="<?php echo esc_url( get_edit_post_link( $post_parent ) ); ?>"><?php echo esc_attr( $title ); ?></a></strong>, <span><?php echo esc_attr__( 'Media uploaded on', 'lurma' ) . ' ' . esc_html( get_the_time( __( 'd/m/Y' ) ) ); ?></span>
			</div>

			<div>
				<a class="hide-if-no-js" onclick="findPosts.open('media[]','<?php echo esc_attr( $post->ID ); ?>'); return false;" href="#the-list"><?php esc_attr_e( 'Re-Attach', 'lurma' ); ?></a>
			</div>

			<div>
				<a href="<?php echo esc_url( $url_unattach ); ?>" title="<?php echo esc_attr__( 'Unattach this media item.', 'lurma' ); ?>"><?php esc_attr_e( 'Unattach', 'lurma' ); ?></a>
			</div>
		<?php

	} else {
		esc_attr_e( '(Unattached)', 'lurma' );
		?>
			<div>
				<a class="hide-if-no-js" onclick="findPosts.open('media[]','<?php echo esc_attr( $post->ID ); ?>'); return false;" href="#the-list"><?php esc_attr_e( 'Attach', 'lurma' ); ?></a>
			</div>
		<?php
	}

}


upgrade-for-unattach-re-attach-media-attachments.php
<?php
/**
 * Plugin Name: Upgrade for Unattach & Re-Attach Media Attachments
 * Plugin URI: https://github.com/bahiirwa/Upgrade-for-Unattach-Re-Attach-Media-Attachments/
 * Description: Safely unattach and Re-attach images and other attachments from within the media library.
 * Version: 1.2.0
 * Author: Laurence Bahiirwa
 * Author URI: https://omukiga.com
 * Requires at least: 4.9
 * Tested up to: 5.4
 * Tags: Attachments, Unattach, Re-Attach, Media Library, Upgrade
 * Text Domain: lurma
 * License: GPLv2
 *
 * @package Lurma
 */

/**
 * Basic Security: Exit plugin if accessed directly.
 */
defined( 'ABSPATH' ) || exit;

/**
 * Define constants
 */
define( 'LURMA_PLUGIN_DIR', plugin_dir_path( __FILE__ ) );

/**
 * The plugin needs to run in two pages that is media and tools.
 */
global $pagenow;

if ( 'upload.php' === $pagenow || 'tools.php' === $pagenow ) {

	require_once LURMA_PLUGIN_DIR . '/includes/functions.php';
	require_once LURMA_PLUGIN_DIR . '/includes/bulk-functions.php';
	require_once LURMA_PLUGIN_DIR . '/includes/manage-columns.php';

	/**
	 * Run Plugin Functions
	 */
	add_action( 'admin_menu', 'lurma_admin_menu' );
	add_filter( 'manage_upload_columns', 'lurma_manage_upload_columns' );
	add_action( 'manage_media_custom_column', 'lurma_manage_media_custom_column', 0, 2 );
	add_action( 'admin_footer', 'lurma_custom_bulk_admin_footer' );
	add_action( 'load-upload.php', 'lurma_custom_bulk_action' );

}

add_action( 'plugins_loaded', 'lurma_load_plugin_textdomain' );

/**
 * Load textdomain.
 *
 * Load gettext translate for text domain.
 *
 * @since 1.0.0
 *
 * @return void
 */
function lurma_load_plugin_textdomain() {
	load_plugin_textdomain( 'lurma' );
}

includes/functions.php
<?php
    if ( ! defined( 'ABSPATH' ) ) {
    	echo 'Hi there! You are up to no good!';
    	exit;
    }
    /*
     * Plugin Functions
     * @package   Lurma
     * @link      https://github.com/bahiirwa/Upgrade-for-Unattach-Re-Attach-Media-Attachments/
     */

    /*
     * Deletes 'post_parent' from an attachment to unattach attachment.
     */
     function lurma_unattach_attachment() {
     	global $wpdb;

     	if (!empty($_REQUEST['post_id'])) {
     		$wpdb->update($wpdb->posts, array('post_parent'=>0),
     		  array('id' => (int)$_REQUEST['post_id'], 'post_type' => 'attachment'));
     	}

     	wp_redirect(admin_url('upload.php'));
     	exit;
     }


     /*
      * Implements action 'admin_menu' to provide a callback for lurma_unattach_attachment().
      */
     function lurma_admin_menu() {
     	if ( current_user_can( 'upload_files' ) ) {
     		//this is hacky but couldn't find the right hook
     		add_submenu_page('tools.php', 'Unattach Media', 'Unattach', 'upload_files', 'unattach', 'lurma_unattach_attachment');
     		remove_submenu_page('tools.php', 'unattach');
     	}
     }


     /*
      * Implements filter 'manage_upload_columns' to replace column 'parent' with
      * out custom column 'extended_parent'.
      */
     function lurma_manage_upload_columns($columns) {
     	unset($columns['parent']);
     	$columns['extended_parent'] = __( 'Parent', 'lurma');
     	return $columns;
     }


     /*
      * Implementes action 'manage_media_custom_column' to add a column into the
      * media page with link Attach, Unattach, Re-Attach.
      */
     function lurma_manage_media_custom_column($column_name, $id) {
     	$post = get_post($id);

     	// Only act on our custom column extended_parent.
     	if ($column_name != 'extended_parent') return;

     	if ($post->post_parent) {
     		if (get_post($post->post_parent)) $title = _draft_or_post_title($post->post_parent);
     		$url_unattach = admin_url('tools.php?page=unattach&noheader=true&post_id=' . $post->ID);
     		?>
         <strong><a href="<?php echo get_edit_post_link( $post->post_parent ); ?>"><?php echo $title ?></a></strong>, <?php echo __('Uploaded on', 'lurma') . ' ' . get_the_time(__('d/m/Y')); ?><br />
         <a class="hide-if-no-js" onclick="findPosts.open('media[]','<?php echo $post->ID ?>'); return false;" href="#the-list"><?php _e('Re-Attach', 'lurma'); ?></a>
         <br />
         <a href="<?php echo esc_url( $url_unattach ); ?>" title="<?php echo __( "Unattach this media item.", 'lurma'); ?>"><?php _e( 'Unattach', 'lurma') ?></a>

        <?php
        } else {
          _e( '(Unattached)', 'lurma' ); ?>
         <br />
         <a class="hide-if-no-js" onclick="findPosts.open('media[]','<?php echo $post->ID ?>'); return false;" href="#the-list"><?php _e('Attach', 'lurma'); ?></a>
         <?php
     	  }
     }

     /**
      * Adds new buld actions 'unattach' and 're-attach' to the media lib.
      *
      * @see http://wordpress.stackexchange.com/questions/29822/custom-bulk-action
     */
     function lurma_custom_bulk_admin_footer() {
     	global $post_type;
     	if( is_admin() ) {
     		?>
          <script type="text/javascript">
     				jQuery(document).ready(function() {
     					$ = jQuery;
     					$('<option>').val('unattach').text('<?php _e('Unattach', 'lurma')?>').appendTo("select[name='action']");
     					$('<option>').val('reattach').text('<?php _e('Re-Attach', 'lurma')?>').appendTo("select[name='action']");
     					$('<option>').val('unattach').text('<?php _e('Unattach', 'lurma')?>').appendTo("select[name='action2']");
     					$('<option>').val('reattach').text('<?php _e('Re-Attach', 'lurma')?>').appendTo("select[name='action2']");

     					$('#doaction, #doaction2').click(function(e){
     					    $('select[name^="action"]').each(function(){
     							if ( $(this).val() == '<?php _e('reattach', 'lurma')?>' ) {
     								e.preventDefault();
     								findPosts.open();
     							}
     						});
     					});
     				});
     			</script>
         <?php
         }
     }

     /*
      * Implements a bulk action for unattaching items in bulk.
      *
      * @see http://wordpress.stackexchange.com/questions/91874/how-to-make-custom-bulk-actions-work-on-the-media-upload-page
      * @see http://www.skyverge.com/blog/add-custom-bulk-action/
      */
     function lurma_custom_bulk_action() {

     	//  ***if($post_type == 'attachment') {  REPLACE WITH:
     	if ( !isset( $_REQUEST['detached'] ) ) {

     		// get the action
     		$wp_list_table = _get_list_table('WP_Media_List_Table');
     		$action = $wp_list_table->current_action();

     		$allowed_actions = array("unattach");
     		if(!in_array($action, $allowed_actions)) return;

     		// security check
     		//  ***check_admin_referer('bulk-posts'); REPLACE WITH:
     		check_admin_referer('bulk-media');

     		// make sure ids are submitted.  depending on the resource type, this may be 'media' or 'ids'
     		if(isset($_REQUEST['media'])) {
     			$post_ids = array_map('intval', $_REQUEST['media']);
     		}

     		if(empty($post_ids)) return;

     		// this is based on wp-admin/edit.php
     		$sendback = remove_query_arg( array('unattached', 'untrashed', 'deleted', 'ids'), wp_get_referer() );
     		if ( ! $sendback )
     			$sendback = admin_url( "upload.php?post_type=$post_type" );

     		$pagenum = $wp_list_table->get_pagenum();
     		$sendback = add_query_arg( 'paged', $pagenum, $sendback );

     		switch($action) {
     			case 'unattach':
     				global $wpdb;

     				// if we set up user permissions/capabilities, the code might look like:
     				//if ( !current_user_can($post_type_object->cap->export_post, $post_id) )
     				//  wp_die( __('You are not allowed to unattach this post.') );
     				if ( !is_admin() )
     					wp_die( __('You are not allowed to unattach this post.', 'lurma') );

     				$unattached = 0;
     				foreach( $post_ids as $post_id ) {
     					// Alter post to unattach media file.
     					if ( $wpdb->update($wpdb->posts, array('post_parent'=>0), array('id' => (int)$post_id, 'post_type' => 'attachment')) === false)
     						wp_die( __('Error unattaching post.', 'lurma') );
     					$unattached++;
     				}

     				$sendback = add_query_arg( array('unattached' => $unattached, 'ids' => join(',', $post_ids) ), $sendback );
     				break;

     			default: return;
     		}

     		$sendback = remove_query_arg( array('action', 'action2', 'tags_input', 'post_author', 'comment_status', 'ping_status', '_status',  'post', 'bulk_edit', 'post_view'), $sendback );

     		wp_redirect($sendback);
     		exit();
     	}
     }

    /*
    * Run Plugin Functions
    */
    add_action( 'admin_menu', 'lurma_admin_menu' );
    add_filter( 'manage_upload_columns', 'lurma_manage_upload_columns') ;
    add_action( 'manage_media_custom_column', 'lurma_manage_media_custom_column', 0, 2 );
    add_action( 'admin_footer', 'lurma_custom_bulk_admin_footer' );
    add_action( 'load-upload.php', 'lurma_custom_bulk_action' );
